design/fake-client.rst

```diff
@@ -0,0 +1,153 @@
+list() which was the earlier implementation. For example, the "status"
+property is no longer set in that case. The second design would automatically
+represent this example of a change.
+
+Design
+------
+
+In the following description, the second design option is used.
+
+There is a fake session object that the user of the zhmcclient package
+will use instead of zhmcclient.Session. The fake session object contains a
+faked HMC that can represent all resources relevant for the zhmcclient.
+There fake session object implements the same API as zhmcclient.Session, so
+that the rest of the zhmcclient classes does not need to be changed.
+The get(), post() and delete() methods of the fake session class will
+be redirected to operate on the faked HMC.
+
+The fake session object will have the ability to be "loaded" with an
+initial state of resources. The resource definitions can be provided as
+a dictionary of nested dictionaries, following this example::
+
+    resources = {
+        'cpcs': [  # name of manager attribute for this resource
+            {
+                'properties': {
+                    # object-id and object-uri are auto-generated
+                    'name': 'cpc_1',
+                    . . .  # more properties
+                },
+                'adapters': [
+                    {
+                        'properties': {
+                            # object-id and object-uri are auto-generated
+                            'name': 'ad_1',
+                            . . .  # more properties
+                        },
+                        'ports': [
+                            {
+                                'properties': {
+                                    # element-id and element-uri are
+                                    # auto-generated
+                                    'name': 'port_1',
+                                    . . .  # more properties
+                                }
+                            },
+                            . . .  # more Ports
+                        ],
+                    },
+                    . . .  # more Adapters
+                ],
+                . . .  # more CPC child resources
+            },
+            . . .  # more CPCs
+        ]
+    }
+
+The properties for object ID ("object-id" or "element-id") and URI
+("object-uri" or "element-uri") are auto-generated, if not provided by
+the user. All other properties will only exist as provided by the user.
+There is no checking for invalid properties.
+
+For all operations against the faked HMC, a successful operation is
+implemented by default. Only if a testcase needs error injection, the user
+needs to deal with setting up the behavior of faked HMC operations.
+
+To support this, the fake session object will have the ability to specify
+non-standard behavior of HMC operations, at the level of the HTTP interactions,
+following this example::
+
+    operations = [
```

Comment:
I guess I would mock hat with python means again... I guess this is similar to the input parameter checking.
It would be really great to have that kind of limits also available in the faked version. Otherwise I have the same risk of doing things wrong like with my faked approach.